### ISSUE TITLE
Image draft not shown on successive forward

### ISSUE THREAD

**Issue by FeuRenard** [WhisperSystems/Signal-Android#5832]
I have:
- [x] searched open and closed issues for duplicates
- [x] read https://github.com/WhisperSystems/Signal-Android/wiki/Submitting-useful-bug-reports

----------------------------------------

### Bug description
The image that should be forwarded is not shown above the compose panel.

### Steps to reproduce
You need two conversations, and ~one~ _both_ of them need~s~ to contain a picture message. ~This one is conversation B.~
- Open conversation A
- Long-press a _picture_ message
- select "Forward message" from the context menu
- select conversation B
- Long-press a media message (optional: with text)
- select "Forward message" from the context menu
- select conversation A
- **critical step**
notice that no picture is shown above the compose panel
optional: The forwarded text is shown in the compose panel. If you press the send button at this point, the message will actually be sent including the previously invisible picture.
- press the back button to leave conversation A
- look at conversation list entry of conversation A and notice that a draft picture was saved
- reenter conversation A and notice that the picture is now shown above the compose panel

**Actual result:** The picture from conversation B is not shown after forwarding it to conversation A.
**Expected result:** The picture from conversation B should be shown above the compose panel of conversation A.

### Device info
**Device:** Fairphone 2
**Android version:** 5.1.1
**Signal version:** 3.21.3

### Device info
**Device:** Sony Xperia U
**Android version:** 4.4.4 (CyanogenMod)
**Signal version:** 3.21.3

**Comment by FeuRenard** [comment]
Now I can describe the issue more general. You have to have an image draft (take a photo in-app, pick one as attachment from your gallery etc.) and then forward an image from this conversation to another conversation. You can forward a conversation image (context menu) or even the draft image itself (via media preview).

### Steps to reproduce (2)

1. Open a conversation.
2. Add an image to your message (e.g. use the in-app camera)
3. Tap this image.
4. Press the forward button
5. Select any conversation
6. **critical step**, see initial post ...

### Explanation
What happens just before the "critical step" is that ConversationActivity is opened and the previous ConversationActivity is reused (see also "launchmode=singleTask"). So you have the old conversation with among other things the old draft image. This situation should be handled by [`onNewIntent(Intent)`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/ConversationActivity.java#L253). In our case two method calls in this method are interesting.


1. [`attachmentManager.clear()`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/ConversationActivity.java#L263), in there [`ViewUtil.fadeOut(attachmentView, 200).addListener...`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/mms/AttachmentManager.java#L96). This seems to happen asynchronously and later the listener is called after the fade out animation is finished. And then after this 200ms and some computing time [`thumbnail.clear()`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/mms/AttachmentManager.java#L99) and [`attachmentView.setVisibility(View.GONE)`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/mms/AttachmentManager.java#L100) are called. Then the thumbnail from the previous conversation is finally cleaned up.
2. [`initializeSecurity(false, false).addListener...`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/ConversationActivity.java#L269) is called. This method does something asynchronously and afterwards [`initializeDraft()`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/ConversationActivity.java#L272) is called via listener. This kicks off some methods - among others [`attachmentManager.setMedia(...)`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/ConversationActivity.java#L1129). This will – again asynchronously – call [`attachmentView.setVisibility(View.VISIBLE)`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/mms/AttachmentManager.java#L214) and [`thumbnail.setImageResource(...)`](https://github.com/WhisperSystems/Signal-Android/blob/b54a271a753fd31dcea77e258e5df5188f1da156/src/org/thoughtcrime/securesms/mms/AttachmentManager.java#L220). Then the new thumbnail should finally be displayed.

### tl;dr
The problem here is, in case you didn't notice yet, a **race condition**. The cleanup/fadeout and the setImage/setVisible are not always called in the right order.

**Comment by FeuRenard** [comment]
This issue should be fixed by cb670d6783e47f8f3c8dc187c8c26a6ee8cbad52. I was not able to reproduce this issue, yet.
